### ISSUE TITLE
Grafting Support

### ISSUE THREAD

**Issue by alainbryden** [alainbryden/bitburner-scripts#387]
One of the next big enhancements that could potentially have a big impact in BN completion times is strategic use of Grafting (available once BN 10 is completed) This is especially useful when there is ample money available (often true when we've been stagnating in a BN for a long time thanks to the stock script)

For example:
- When the only remaining requirement for Daedalus is hack, grafting the remaining aug(s) that gives the biggest hack boost might be faster than training and/or grinding reputation with the factions that would provide these augs and installing.
- In Bitnodes where reputation gain is severely penalized, so grinding reputation is not worthwhile
- In (if I recall correctly) the gangs bitnode when you can only get augmentations from your gang, but certain augs are not available? (this may no longer be the case)

I also have been seeing work-for-faction.js spend quite a lot of money and time (sometimes hours) trying to study/train a certain stat. (The current heuristic for determining whether training/studying is worthwhile is not great). If we can improve this heuristic to use the actual rep gain rate and determine how long we'd have to train to achieve the desired stat, we can compare that to the time to Graft 1 or 2 augs instead.

Note that everything gets more complicated because we have to take into account stat reductions from entropy virus until we can afford "violet Congruity Implant" to cure it. This can have complex interactions, such as breaking daemon.js' assumptions about being able to hack targets when our hack level suddenly decreases. While it's sub-optimal, we may wish to wait on doing *any* grafting until we have plenty more than the 150t required to first install this (and make sure spending this money wouldn't be too much of a setback vs. just installing all the available augs.)

**Comment by alainbryden** [comment]
Even before adding general grafting support, I may want to add some special case logic for BN8 to set up a sequence of 30 augmentations to graft, focusing on the very cheapest ones (to boost aug count without eating too much of our investment income) as well as ones that boost hack level / exp (to unlock factions - especially daedalus)

**Comment by alainbryden** [comment]
#166 might be an early duplicate of this
